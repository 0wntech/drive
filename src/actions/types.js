export const LOGIN = 'login';
export const LOGIN_SUCCESS = 'login_success';
export const LOGIN_FAIL = 'login_fail';

export const FETCH_USER = 'fetch_user';
export const FETCH_USER_SUCCESS = 'fetch_user_success';
export const FETCH_USER_FAIL = 'fetch_user_fail';

export const SET_WEBID = 'set_webid';
export const SET_CURRENT_PATH = 'set_current_path';

export const LOAD_CURRENT_ITEMS = 'load_current_items';
export const SET_CURRENT_ITEMS = 'set_current_items';

export const FETCH_FRIENDS = 'fetch_friends';
export const FETCH_FRIENDS_SUCCESS = 'fetch_friends_success';
export const FETCH_FRIENDS_FAIL = 'fetch_friends_fail';

export const FETCH_CURRENT_ITEM = 'fetch_current_item';
export const FETCH_CURRENT_ITEM_SUCCESS = 'fetch_current_item_success';
export const FETCH_CURRENT_ITEM_FAIL = 'fetch_current_item__fail';

export const GET_SESSION = 'get_session';
export const GET_SESSION_SUCCESS = 'get_session_success';
export const GET_SESSION_FAIL = 'get_session_fail';

export const FETCH_NOTIFICATIONS = 'fetch_notifications';
export const FETCH_NOTIFICATIONS_SUCCESS = 'fetch_notifications_success';
export const FETCH_NOTIFICATIONS_FAILURE = 'fetch_notifications_failure';

export const SET_SELECTION = 'set_selection';

export const SEND_NOTIFICATION = 'send_notification';
export const SEND_NOTIFICATION_SUCCESS = 'send_notification_success';
export const SEND_NOTIFICATION_FAILURE = 'send_notification_failure';

export const FETCH_IDPS = 'fetch_idps';
export const FETCH_IDPS_SUCCESS = 'fetch_idps_success';
export const FETCH_IDPS_FAILED = 'fetch_idps_failure';

export const DELETE_ITEMS = 'delete_items';
export const DELETE_ITEMS_SUCCESS = 'delete_items_success';
export const DELETE_ITEMS_FAILURE = 'delete_items_failure';

export const COPY_ITEMS = 'copy_items';
export const PASTE_ITEMS = 'paste_items';
export const PASTE_ITEMS_SUCCESS = 'paste_items_success';
export const PASTE_ITEMS_FAILURE = 'paste_items_failure';

export const FETCH_CONTACTS = 'fetch_contacts';
export const FETCH_CONTACTS_SUCCESS = 'fetch_contacts_success';
export const FETCH_CONTACTS_FAILURE = 'fetch_contacts_failure';

export const ADD_CONTACT = 'add_contact';
export const ADD_CONTACT_SUCCESS = 'add_contact_success';
export const ADD_CONTACT_FAILURE = 'add_contact_failure';

export const REMOVE_CONTACT = 'remove_contact';
export const REMOVE_CONTACT_SUCCESS = 'remove_contact_success';
export const REMOVE_CONTACT_FAILURE = 'remove_contact_failure';

export const UPDATE_PROFILE = 'delete_items';
export const UPDATE_PROFILE_SUCCESS = 'update_profile_success';
export const UPDATE_PROFILE_FAILURE = 'update_profile_failure';

export const CHANGE_PROFILE_PHOTO = 'change_profile_photo';
export const CHANGE_PROFILE_PHOTO_SUCCESS = 'change_profile_photo_success';
export const CHANGE_PROFILE_PHOTO_FAILURE = 'change_profile_photo_failure';

export const SET_CURRENT_CONTACT = 'set_current_contact';

export const RENAME_ITEM = 'rename_item';
export const RENAME_ITEM_SUCCESS = 'rename_item_success';
export const RENAME_ITEM_FAILURE = 'rename_item_failure';

export const SEARCH_CONTACT = 'search_contact';
export const SEARCH_CONTACT_SUCCESS = 'search_contact_success';
export const SEARCH_CONTACT_FAILURE = 'search_contact_failure';
