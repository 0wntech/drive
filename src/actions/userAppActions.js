import { FETCH_APPS, FETCH_APPS_SUCCESS, FETCH_APPS_FAILURE } from './types';
import User from 'ownuser';

// setApp([urls of apps])
// { getApps, setApps }
export const fetchApps = (webId) => {
    return (dispatch) => {
        dispatch({ type: FETCH_APPS });
        const user = new User(webId);
        user.getApps()
            .then((apps) => {
                console.log(apps);
                dispatch({ type: FETCH_APPS_SUCCESS, payload: apps });
            })
            .catch((error) =>
                dispatch({ type: FETCH_APPS_FAILURE, payload: error })
            );
    };
};
