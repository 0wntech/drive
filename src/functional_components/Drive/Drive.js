import React, { useEffect } from 'react';
import url from 'url';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import styles from './Drive.module.css';
import Breadcrumbs from '../Breadcrumbs/Breadcrumbs';
import { ItemList } from '../ItemList';
import fileUtils from '../../utils/fileUtils';
import { getBreadcrumbsFromUrl } from '../../utils/url';
import folder from '../../assets/icons/Folder.png';
import fileIcon from '../../assets/icons/File.png';
import { Layout } from '../Layout';
import {
    setCurrentPath,
    setSelection,
    sendNotification,
    fetchCurrentItem,
} from '../../actions/appActions';
import { ClassicSpinner } from 'react-spinners-kit';
import ToolbarButtons from '../ToolbarButtons/ToolbarButtons';
import { isCmdPressed } from '../../utils/helper';
import Windows from '../Windows/Windows';
import { CursorMenu } from '../DriveContextMenu/DriveContextMenu';

const Drive = ({
    selectedItems,
    currentItem,
    currentPath,
    loadCurrentItem,
    webId,
    setCurrentPath,
    setSelection,
    fetchCurrentItem,
    history,
    loadDeletion,
    loadPaste,
}) => {
    useEffect(() => {
        if (!currentPath && !loadCurrentItem && webId) {
            currentPath = webId.replace('profile/card#me', '');
            setCurrentPath(currentPath, true);
        } else if (currentPath && currentItem && currentItem.body) {
            currentPath =
                currentPath.substr(0, currentPath.lastIndexOf('/')) + '/';
            setCurrentPath(currentPath);
        }
    });

    // Event Handlers
    const loadFile = (url, event = {}) => {
        if (url.endsWith('/')) {
            url = url.substr(0, url.lastIndexOf('/'));
        }
        if (isCmdPressed(event) && selectedItems.includes(url) === false) {
            const newSelection = [...selectedItems];
            newSelection.push(url);
            setSelection(newSelection);
        } else if (isCmdPressed(event) && selectedItems.includes(url)) {
            const newSelection = selectedItems.filter((item) => item !== url);
            setSelection(newSelection);
        } else {
            history.push(`/file?f=${url}`);
        }
    };

    const loadFolder = (path, event = {}) => {
        if (loadCurrentItem) return;
        if (isCmdPressed(event) && selectedItems.includes(path)) {
            const newSelection = selectedItems.filter((item) => item !== path);
            setSelection(newSelection);
        } else if (isCmdPressed(event) && !selectedItems.includes(path)) {
            const newSelection = [...selectedItems, path];
            setSelection(newSelection);
        } else {
            setCurrentPath(path);
            fetchCurrentItem(path, true);
        }
    };

    const uploadFile = (e) => {
        const filePath =
            e.target.files && e.target.files.length ? e.target.files[0] : null;
        if (filePath) {
            fileUtils.uploadFile(filePath, currentPath, () => {
                fetchCurrentItem(currentPath);
            });
        }
    };

    const clearSelection = (e) => {
        if (
            typeof e.target.className === 'string' &&
            !e.target.className.includes('Item_') &&
            !e.target.className.includes('File_') &&
            e.target.className !== ''
        ) {
            console.log(e.target);
            console.log('Emptying selection');
            setSelection([]);
        }
    };

    const downloadItems = () => {
        selectedItems.forEach((item) => {
            const download =
                webId.replace('profile/card#me', 'download?path=') +
                url.parse(item).pathname;
            window.open(download.replace(/\/+$/, ''));
        });
    };

    const uploadCurrentItem = (e) => {
        const files = e.target.files;
        if (files && files.length) {
            for (let file = 0; file < files.length; file++) {
                fileUtils
                    .uploadCurrentItemOrFile(
                        files[file],
                        currentPath +
                            encodeURIComponent(files[file].webkitRelativePath)
                    )
                    .then((response) => {
                        console.log(file, response);
                        if (file === files.length - 1) {
                            fetchCurrentItem(currentPath);
                        }
                    });
            }
        }
    };

    // toolbar fragments
    const toolbarRight = (
        <ToolbarButtons
            onFolderUpload={uploadCurrentItem}
            onDownload={downloadItems}
            uploadFile={uploadFile}
            onDelete={() => {
                const deletable = addForDelete();
                if (deletable) openConsentWindow();
            }}
        />
    );

    const toolbarLeft = webId ? (
        <div className={styles.breadcrumbsContainer}>
            <Breadcrumbs
                onClick={(path) => {
                    console.log(path);
                    setCurrentPath(path);
                }}
                breadcrumbs={
                    currentPath ? getBreadcrumbsFromUrl(currentPath) : null
                }
                webId={webId}
            />
        </div>
    ) : null;

    if ((loadCurrentItem, loadDeletion, loadPaste)) {
        return (
            <div className={styles.spinner}>
                <ClassicSpinner
                    size={30}
                    color="#686769"
                    loading={(loadCurrentItem, loadDeletion, loadPaste)}
                />
            </div>
        );
    } else {
        return (
            <Layout
                toolbarChildrenLeft={toolbarLeft}
                toolbarChildrenRight={toolbarRight}
                className={styles.grid}
                label="Drive"
                onClick={clearSelection}
            >
                <CursorMenu
                    className={styles.mainArea}
                    drive
                    id="drive contextmenu"
                >
                    <div className={styles.container}>
                        <Windows />
                        {currentItem &&
                        (currentItem.folders || currentItem.files) ? (
                            <div>
                                {currentItem.folders.length > 0 ? (
                                    <ItemList
                                        selectedItems={selectedItems}
                                        items={currentItem.folders}
                                        currPath={currentPath}
                                        image={folder}
                                        onItemClick={loadFolder}
                                    />
                                ) : null}
                                <ItemList
                                    selectedItems={selectedItems}
                                    isFile
                                    items={currentItem.files}
                                    currPath={currentPath}
                                    image={fileIcon}
                                    onItemClick={loadFile}
                                />
                            </div>
                        ) : (
                            undefined
                        )}
                    </div>
                </CursorMenu>
            </Layout>
        );
    }
};

const mapStateToProps = (state) => {
    return {
        currentItem: state.app.currentItem,
        currentPath: state.app.currentPath,
        selectedItems: state.app.selectedItems,
        webId: state.user.webId,
        loadCurrentItem: state.app.loadCurrentItem,
        loadDeletion: state.app.loadDeletion,
        clipboard: state.app.clipboard,
        loadPaste: state.app.loadPaste,
        loadCurrentItem: state.app.loadCurrentItem,
    };
};

export default withRouter(
    connect(mapStateToProps, {
        setCurrentPath,
        sendNotification,
        fetchCurrentItem,
        setSelection,
    })(Drive)
);
