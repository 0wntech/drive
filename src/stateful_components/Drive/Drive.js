import React, { Fragment } from 'react';
import rdf from 'rdflib';
import auth from 'solid-auth-client';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import styles from './Drive.module.css';
import Breadcrumbs from '../../functional_components/Breadcrumbs/Breadcrumbs';
import { ItemList } from '../../functional_components/ItemList';
import fileUtils from '../../utils/fileUtils';
import { getBreadcrumbsFromUrl } from '../../utils/url';
import folder from '../../assets/icons/Folder.png';
import fileIcon from '../../assets/icons/File.png';
import { InputWindow } from '../../functional_components/InputWindow';
import { DeleteWindow } from '../../functional_components/DeleteWindow';
import {
    setCurrentPath,
    setSelection,
    sendNotification,
    fetchCurrentItems,
    deleteItems,
} from '../../actions/UserActions';
import { ClassicSpinner } from 'react-spinners-kit';
import ToolbarButtons from '../../functional_components/ToolbarButtons';
import { isCmdPressed } from '../../utils/helper';
import { MenuProvider, Menu, Item } from 'react-contexify';
import classNames from 'classnames';
import url from 'url';
const ns = require('solid-namespace')(rdf);

class Drive extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            currPath: undefined,
            file: undefined,
            image: undefined,
            folders: undefined,
            isCreateFolderVisible: false,
            isConsentWindowVisible: false,
            isCreateFileVisible: false,
            files: undefined,
        };

        this.createFolder = this.createFolder.bind(this);
        this.createFile = this.createFile.bind(this);
        this.followPath = this.followPath.bind(this);
        this.uploadFolder = this.uploadFolder.bind(this);
        this.uploadFile = this.uploadFile.bind(this);
        this.loadFile = this.loadFile.bind(this);
        this.loadCurrentFolder = this.loadCurrentFolder.bind(this);
        this.clearSelection = this.clearSelection.bind(this);
        this.openCreateFolderWindow = this.openCreateFolderWindow.bind(this);
        this.closeCreateFolderWindow = this.closeCreateFolderWindow.bind(this);
        this.openConsentWindow = this.openConsentWindow.bind(this);
        this.closeConsentWindow = this.closeConsentWindow.bind(this);
        this.openCreateFileWindow = this.openCreateFileWindow.bind(this);
        this.closeCreateFileWindow = this.closeCreateFileWindow.bind(this);
    }

    sortContainments(urls) {
        const folders = [];
        const files = [];
        urls.forEach((url) => {
            if (url.value[url.value.length - 1] === '/') {
                const urlFragments = url.value.split('/');
                const folderUrl = urlFragments[urlFragments.length - 2];
                folders.push(decodeURIComponent(folderUrl));
            } else {
                const urlFragments = url.value.split('/');
                const fileUrl = urlFragments[urlFragments.length - 1];
                files.push(decodeURIComponent(fileUrl));
            }
        });
        return [files, folders];
    }

    loadFolder(url) {
        const store = rdf.graph();
        const fetcher = new rdf.Fetcher(store);

        return fetcher.load(url).then(() => {
            const containments = store.each(
                rdf.sym(url),
                ns.ldp('contains'),
                null
            );

            return this.sortContainments(containments);
        });
    }

    loadCurrentFolder(path, newBreadcrumbs) {
        const currPath = path
            ? path
            : 'https://' + this.props.webId.split('/')[2] + '/';
        Promise.resolve(this.loadFolder(currPath, newBreadcrumbs)).then(
            (sortedContainments) => {
                this.setState({
                    folders: sortedContainments[1],
                    files: sortedContainments[0],
                    currPath: currPath,
                    breadcrumbs: newBreadcrumbs,
                    file: undefined,
                    image: undefined,
                    selectedItems: [],
                });
            }
        );
    }

    loadFile(url, event = {}) {
        const { selectedItems, setSelection } = this.props;
        if (
            event.metaKey ||
            (isCmdPressed(event) && selectedItems.includes(url) === false)
        ) {
            const newSelection = [...selectedItems];
            newSelection.push(url);
            setSelection(newSelection);
        } else if (
            event.metaKey ||
            (isCmdPressed(event) && selectedItems.includes(url))
        ) {
            const newSelection = selectedItems.filter((item) => item != url);
            setSelection(newSelection);
        } else {
            const newBreadCrumbs = getBreadcrumbsFromUrl(url);

            const contentType = fileUtils.getContentType(url);
            if (contentType === 'image') {
                this.setState({
                    file: url,
                    image: url,
                    currPath: url,
                    breadcrumbs: newBreadCrumbs,
                    selectedItems: [],
                });
                return;
            }

            auth.fetch(url).then((response) => {
                response.text().then((text) => {
                    this.setState({
                        file: text,
                        currPath: url,
                        breadcrumbs: newBreadCrumbs,
                        selectedItems: [],
                    });
                });
            });
        }
    }

    followPath(path, event = {}) {
        const { selectedItems, setCurrentPath, setSelection } = this.props;
        if (isCmdPressed(event) && selectedItems.includes(path)) {
            const newSelection = selectedItems.filter((item) => item != path);
            setSelection(newSelection);
        } else if (isCmdPressed(event) && !selectedItems.includes(path)) {
            const newSelection = [...selectedItems, path];
            setSelection(newSelection);
        } else {
            setCurrentPath(path);
        }
    }

    uploadFile(e) {
        console.log(e);
        const currPath = this.props.currentPath;
        const filePath = e.target.files[0];

        fileUtils.uploadFile(filePath, currPath).then(() => {
            this.setCurrentPath(this.props.currentPath);
        });
    }

    clearSelection(e) {
        if (
            e.target.nodeName !== 'IMG' &&
            e.target.nodeName !== 'P' &&
            e.target.innerHTML !== 'Delete'
        ) {
            this.setState({
                selectedItems: [],
            });
        }
    }

    createFolder(folderAddress) {
        const request = {
            method: 'POST',
            headers: {
                slug: folderAddress,
                link: '<http://www.w3.org/ns/ldp#BasicContainer>; rel="type"',
                'Content-Type': 'text/turtle',
            },
        };

        auth.fetch(this.props.currentPath, request).then(() => {
            this.props.setCurrentPath(this.props.currentPath);
        });
    }

    createFile(folderAddress) {
        const request = {
            method: 'POST',
            headers: {
                slug: folderAddress,
                link: '<http://www.w3.org/ns/ldp#Resource>; rel="type"',
                contentType: 'text/turtle',
            },
        };

        auth.fetch(this.props.currentPath, request).then(() => {
            this.props.setCurrentPath(this.props.currentPath);
        });
    }

    componentDidMount() {
        const { currentItems, currentPath, loadCurrentItems } = this.props;

        if (!currentItems && !currentPath && !loadCurrentItems) {
            console.log('fetching files');
            console.log(
                'current path + current items: ',
                currentPath,
                currentItems
            );
            fetchCurrentItems(currentPath);
        }

        // if (webId) {
        //     const object = webId.replace('/card#me', '');
        //     console.log('sending notification...')
        //     sendNotification({ actor: webId, object: object, target: webId });
        // }
        // try {
        //     if (!JSON.parse(localStorage.getItem('appState')).currPath) {
        //         this.loadCurrentFolder(this.state.currPath, ['/']);
        //     } else {
        //         console.log('Using cached state...');
        //         this.loadCurrentFolder(
        //             JSON.parse(localStorage.getItem('appState')).currPath,
        //             JSON.parse(localStorage.getItem('appState')).breadcrumbs
        //         );
        //     }
        // } catch (e) {
        //     console.log(e);
        // }
    }

    uploadFolder(e) {
        const files = e.target.files;
        for (let file = 0; file < files.length; file++) {
            fileUtils
                .uploadFolderOrFile(
                    files[file],
                    this.props.currentPath +
                        encodeURIComponent(files[file].webkitRelativePath)
                )
                .then((response) => {
                    console.log(file, response);
                    if (file === files.length - 1) {
                        this.loadCurrentFolder(
                            this.props.currentPath,
                            getBreadcrumbsFromUrl(this.props.currentPath)
                        );
                    }
                });
        }
    }

    closeCreateFolderWindow() {
        this.setState({
            isCreateFolderVisible: false,
        });
    }

    openCreateFolderWindow() {
        this.setState({
            isCreateFolderVisible: true,
        });
    }

    closeConsentWindow() {
        this.setState({
            isConsentWindowVisible: false,
        });
    }

    openConsentWindow(item) {
        const { selectedItems, webId, setSelection } = this.props;
        const newSelection = [...selectedItems];
        if (
            item &&
            url.parse(item).hostname !== url.parse(webId).hostname &&
            !selectedItems.includes(item)
        ) {
            console.log(url.parse(webId).hostname, item);
            newSelection.push(item);
            setSelection(newSelection);
        }
        if (selectedItems.length !== 0 || newSelection.length !== 0) {
            this.setState({
                isConsentWindowVisible: true,
            });
        }
    }

    closeCreateFileWindow() {
        this.setState({
            isCreateFileVisible: false,
        });
    }

    openCreateFileWindow() {
        this.setState({
            isCreateFileVisible: true,
        });
    }

    componentWillUnmount() {
        // console.log('Caching state...');
        // localStorage.setItem('appState', JSON.stringify(this.state));
    }

    render() {
        const {
            selectedItems,
            currentItems,
            currentPath,
            deleteItems,
            loadCurrentItems,
            webId,
            setCurrentPath,
            loadDeletion,
        } = this.props;

        const {
            isConsentWindowVisible,
            isCreateFileVisible,
            isCreateFolderVisible,
        } = this.state;

        const CONTEXTMENU_OPTIONS = [
            {
                label: 'Info',
                onClick: (item) => fileUtils.getInfo(item),
                disabled: false,
            },
            {
                label: 'Copy*',
                onClick: (item) => fileUtils.getInfo(item),
                disabled: true,
            },
            {
                label: 'Paste*',
                onClick: (item) => fileUtils.getInfo(item),
                disabled: true,
            },
            {
                label: 'Rename',
                onClick: (item) => fileUtils.renameItem(item),
                disabled: false,
            },
            {
                label: 'Manage Access',
                onClick: (item) => fileUtils.changeAccess(item),
                disabled: false,
            },
            {
                label: 'Share*',
                onClick: (item) => fileUtils.changeAccess(item),
                disabled: true,
            },
            {
                label: 'Create Folder',
                onClick: () => this.openCreateFolderWindow(),
            },
            {
                label: 'Delete',
                onClick: (item) => this.openConsentWindow(item),
                disabled: false,
            },
        ];
        const fileMarkup = this.state.file ? (
            <div className={styles.renderedFile}>
                {this.state.image ? (
                    <img src={this.state.image} alt="requested file"></img>
                ) : (
                    this.state.file
                )}
            </div>
        ) : (
            undefined
        );

        const toolbar = (
            <div className={styles.toolbarArea}>
                {webId ? (
                    <Breadcrumbs
                        onClick={setCurrentPath}
                        breadcrumbs={
                            currentPath
                                ? getBreadcrumbsFromUrl(currentPath)
                                : null
                        }
                        webId={webId}
                    />
                ) : null}

                <ToolbarButtons
                    onFileCreation={this.openCreateFileWindow}
                    onFolderCreation={this.openCreateFolderWindow}
                    onFolderUpload={this.uploadFolder}
                    uploadFile={this.uploadFile}
                    onDelete={() => {
                        this.openConsentWindow();
                    }}
                />
            </div>
        );

        const windows = (
            <Fragment>
                <DeleteWindow
                    windowName="Delete File"
                    selectedItems={selectedItems}
                    info={
                        selectedItems.length > 1
                            ? 'Do you really want to delete these items?'
                            : 'Do you really want to delete this item?'
                    }
                    onSubmit={(selectedItems) => {
                        deleteItems(selectedItems, currentPath);
                    }}
                    className={
                        isConsentWindowVisible ? styles.visible : styles.hidden
                    }
                    onClose={this.closeConsentWindow}
                ></DeleteWindow>
                <InputWindow
                    windowName="Create Folder"
                    info=""
                    onSubmit={(value) => this.createFolder(value)}
                    className={
                        isCreateFolderVisible ? styles.visible : styles.hidden
                    }
                    onClose={this.closeCreateFolderWindow}
                    placeholder={'Untitled'}
                />
                <InputWindow
                    windowName="Create File"
                    info=""
                    onSubmit={(value) => this.createFile(value)}
                    className={
                        isCreateFileVisible ? styles.visible : styles.hidden
                    }
                    onClose={this.closeCreateFileWindow}
                    placeholder={'Untitled'}
                />
            </Fragment>
        );

        if ((loadCurrentItems, loadDeletion)) {
            return (
                <div className={styles.spinner}>
                    <ClassicSpinner
                        size={30}
                        color="#686769"
                        loading={(loadCurrentItems, loadDeletion)}
                    />
                </div>
            );
        } else {
            return (
                <div
                    className={styles.grid}
                    style={{ height: '100%' }}
                    onClick={this.clearSelection}
                >
                    {toolbar}
                    <MenuProvider
                        className={styles.mainArea}
                        id="drive contextmenu"
                    >
                        {fileMarkup ? (
                            <div className={styles.container}>{fileMarkup}</div>
                        ) : (
                            <div className={styles.container}>
                                {windows}
                                {currentItems ? (
                                    <div className={styles.contentWrapper}>
                                        {/* <ContactSidebar /> */}
                                        <div className={styles.header}>
                                            Folders
                                        </div>
                                        <ItemList
                                            selectedItems={selectedItems}
                                            items={currentItems.folders}
                                            currPath={currentPath}
                                            image={folder}
                                            onItemClick={this.followPath}
                                            contextMenuOptions={
                                                CONTEXTMENU_OPTIONS
                                            }
                                        />
                                        <div className={styles.header}>
                                            Files
                                        </div>
                                        <ItemList
                                            selectedItems={selectedItems}
                                            isFile
                                            items={currentItems.files}
                                            currPath={currentPath}
                                            image={fileIcon}
                                            onItemClick={this.loadFile}
                                            contextMenuOptions={
                                                CONTEXTMENU_OPTIONS
                                            }
                                        />
                                    </div>
                                ) : (
                                    undefined
                                )}
                            </div>
                        )}
                    </MenuProvider>
                    <Menu
                        className={styles.contextMenu}
                        id={'drive contextmenu'}
                    >
                        {CONTEXTMENU_OPTIONS &&
                            CONTEXTMENU_OPTIONS.map((option, index) => (
                                <Item
                                    disabled={option.disabled}
                                    key={index + option.label}
                                    onClick={
                                        !option.disabled
                                            ? () => {
                                                  option.onClick(currentPath);
                                              }
                                            : undefined
                                    }
                                    className={classNames(styles.contextItem, {
                                        [styles.disabled]: option.disabled,
                                    })}
                                >
                                    <div>{option.label}</div>
                                </Item>
                            ))}
                    </Menu>
                </div>
            );
        }
    }
}

const mapStateToProps = (state) => {
    return {
        currentItems: state.app.currentItems,
        currentPath: state.app.currentPath,
        selectedItems: state.app.selectedItems,
        webId: state.app.webId,
        loadCurrentItems: state.app.loadCurrentItems,
        loadDeletion: state.app.loadDeletion,
    };
};

export default withRouter(
    connect(
        mapStateToProps,
        {
            setCurrentPath,
            sendNotification,
            fetchCurrentItems,
            setSelection,
            deleteItems,
        }
    )(Drive)
);
