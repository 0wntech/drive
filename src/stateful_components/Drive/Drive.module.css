.spinner {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translateX(-50%) translateY(-50%);
}

.fileMarkup li {
    list-style: none;
}

.hidden {
    display: none;
}

.visible {
    display: initial;
}

.toolbarArea {
    grid-area: toolbar;
    background-color: #e2e2e6;
    border-top-left-radius: 20px;
    border-top-right-radius: 20px;
    width: 100%;
    padding: 0 15px 0 15px;
    display: flex;
    justify-content: space-between;
}

.mainArea {
    grid-area: main;
    overflow-y: auto;
}

.mainArea::-webkit-scrollbar {
    display: none;
}

.grid {
    display: grid;
    grid-template-rows: 40px 23fr;
    grid-template-columns: 1fr 10fr 1fr;
    grid-template-areas: 'toolbar toolbar toolbar' '. main .';
    background-color: #ffffff;
    border-radius: 20px;
}

.contentWrapper {
    overflow: hidden;
    grid-area: content;
}

.header {
    font-size: 32px;
    line-height: 40px;
    letter-spacing: 0.05em;
    font-family: Source Sans Pro;
    margin: 30px 0 20px 0;
}

.contextMenu {
    background-color: white;
    border-radius: 5px;
    border: 0.5px solid #ddd;
    z-index: 2;
    width: 130px;
    padding: 0 !important;
}

.contextItem div {
    color: #000000;
}

.contextItem:hover div {
    background-color: #f2f2f2 !important;
}

.disabled div {
    color: #8c8c8c;
    cursor: default;
}
