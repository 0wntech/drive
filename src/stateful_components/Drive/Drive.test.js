import React from 'react';
import Drive from './Drive.js';
import ShallowRenderer from 'react-test-renderer/shallow';
import appReducer from '../../reducers/appReducer.js';
import { FETCH_CURRENT_ITEM_SUCCESS } from '../../actions/types';

describe('Drive', () => {
    const currentItem = {
        files: ['haha.jpg'],
        folders: ['hehe', 'hihi'],
    };

    describe('Drive Redux', () => {
        const initialState = appReducer(undefined, {});
        it('receives current folder', () => {
            const result = appReducer(initialState, {
                type: FETCH_CURRENT_ITEM_SUCCESS,
                payload: currentItem,
            });

            expect(result).toStrictEqual({
                ...initialState,
                loadCurrentItem: false,
                currentItem: currentItem,
            });
        });
    });

    describe('Drive Snapshot', () => {
        it('matches the snapshot', () => {
            const result = new ShallowRenderer().render(
                <Drive.WrappedComponent />
            );
            expect(result).toMatchSnapshot();
        });
    });
});
