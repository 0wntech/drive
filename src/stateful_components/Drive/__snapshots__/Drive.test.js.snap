// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drive Drive Snapshot matches the snapshot 1`] = `
<Layout
  className="grid"
  label="Drive"
  onClick={[Function]}
  toolbarChildrenLeft={null}
  toolbarChildrenRight={
    <ToolbarButtons
      onDelete={[Function]}
      onDownload={[Function]}
      onFolderUpload={[Function]}
      uploadFile={[Function]}
    />
  }
>
  <MenuProvider
    className="mainArea"
    component="div"
    event="onContextMenu"
    id="drive contextmenu"
    storeRef={true}
  >
    <div
      className="container"
    >
      <Memo(Connect(Windows)) />
    </div>
  </MenuProvider>
  <Menu
    className="contextMenu"
    id="drive contextmenu"
  >
    <Item
      className="contextItem"
      disabled={false}
      onClick={[Function]}
    >
      <div>
        Info
      </div>
    </Item>
    <Item
      className="contextItem"
      disabled={false}
      onClick={[Function]}
    >
      <div>
        Copy
      </div>
    </Item>
    <Item
      className="contextItem"
      disabled={false}
      onClick={[Function]}
    >
      <div>
        Paste
      </div>
    </Item>
    <Item
      className="contextItem"
      disabled={false}
      onClick={[Function]}
    >
      <div>
        Manage Access
      </div>
    </Item>
    <Item
      className="contextItem disabled"
      disabled={true}
      onClick={[Function]}
    >
      <div>
        Share*
      </div>
    </Item>
    <Item
      className="contextItem"
      disabled={false}
      onClick={[Function]}
    >
      <div>
        Create Folder
      </div>
    </Item>
    <Item
      className="contextItem"
      disabled={false}
      onClick={[Function]}
    >
      <div>
        Create File
      </div>
    </Item>
    <Item
      className="contextItem"
      disabled={false}
      onClick={[Function]}
    >
      <div>
        Delete
      </div>
    </Item>
  </Menu>
</Layout>
`;
